## Classification using Convolutional Neural Network

Convolutional Neural Network (CNN) has revolutionized the computer vision space. As one of the recent breakthrough in deep learning, CNN has become the state-of-the-art in computer vision technique. 

Here, CNN algorithm will be applied using Keras---a Python API for deep learning that uses TensorFlow as backend. The algorithm will be built to classify hand-written digits on MNIST Dataset. However, we will first apply the simple fully-connected layers before investigating CNN.

## Fully-connected

In [1]:
# Import libraries

from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Conv2D, MaxPool2D, Flatten
from keras.utils import np_utils
from sklearn.metrics import accuracy_score

In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Flattening images from 28x28 pixels to 1D
X_train = X_train.reshape(60000, 784).astype('float32')
X_test = X_test.reshape(10000, 784).astype('float32')

# Normalize the dataset to help with the stability of training
X_train /= 255.
X_test /= 255.

# One-hot encoding for the output
n_classes = 10
print("Shape before one-hot encoding: ", y_train.shape)
Y_train = np_utils.to_categorical(y_train, n_classes)
Y_test = np_utils.to_categorical(y_test, n_classes)
print("Shape after one-hot encoding: ", Y_train.shape)

Shape before one-hot encoding:  (60000,)
Shape after one-hot encoding:  (60000, 10)


In [3]:
# Digit 5 is represented as an array of
print(y_train[0], ": ", Y_train[0])

5 :  [0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]


In [4]:
# Building a linear stack of layers with the sequential model
model = Sequential()
model.add(Dense(100, input_shape=(784,), activation='relu')) # hidden layer 1
model.add(Dense(100, input_shape=(100,), activation='relu')) # hidden layer 2
model.add(Dense(100, input_shape=(100,), activation='relu')) # hidden layer 3
model.add(Dense(100, input_shape=(100,), activation='relu')) # hidden layer 4
model.add(Dense(10, activation='softmax')) # output layer
model.summary() # print model summary

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 100)               78500     
                                                                 
 dense_1 (Dense)             (None, 100)               10100     
                                                                 
 dense_2 (Dense)             (None, 100)               10100     
                                                                 
 dense_3 (Dense)             (None, 100)               10100     
                                                                 
 dense_4 (Dense)             (None, 10)                1010      
                                                                 
Total params: 109,810
Trainable params: 109,810
Non-trainable params: 0
_________________________________________________________________


In [5]:
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam') # construct loss function
model.fit(X_train, Y_train, batch_size=128, epochs=10, validation_data=(X_test, Y_test)) # training

Epoch 1/10
469/469 [==============================] - 1s 2ms/step - loss: 0.3302 - accuracy: 0.9046 - val_loss: 0.1363 - val_accuracy: 0.9588
Epoch 2/10
469/469 [==============================] - 1s 2ms/step - loss: 0.1234 - accuracy: 0.9625 - val_loss: 0.1102 - val_accuracy: 0.9656
Epoch 3/10
469/469 [==============================] - 1s 2ms/step - loss: 0.0856 - accuracy: 0.9734 - val_loss: 0.0997 - val_accuracy: 0.9689
Epoch 4/10
469/469 [==============================] - 1s 2ms/step - loss: 0.0662 - accuracy: 0.9791 - val_loss: 0.0858 - val_accuracy: 0.9738
Epoch 5/10
469/469 [==============================] - 1s 2ms/step - loss: 0.0518 - accuracy: 0.9830 - val_loss: 0.1103 - val_accuracy: 0.9661
Epoch 6/10
469/469 [==============================] - 1s 2ms/step - loss: 0.0436 - accuracy: 0.9862 - val_loss: 0.0825 - val_accuracy: 0.9744
Epoch 7/10
469/469 [==============================] - 1s 2ms/step - loss: 0.0381 - accuracy: 0.9876 - val_loss: 0.0835 - val_accuracy: 0.9738
Epoch 

#### Validation accuracy of 97% is the best we can do with fully-connected layers.

## Convolutional Neural Network

In [6]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# building the input vector from the 28x28 pixels
X_train = X_train.reshape(X_train.shape[0], 28, 28, 1).astype('float32')
X_test = X_test.reshape(X_test.shape[0], 28, 28, 1).astype('float32')

# Normalize the dataset to help with the stability of training
X_train /= 255.
X_test /= 255.

# One-hot encoding for the output
n_classes = 10
print("Shape before one-hot encoding: ", y_train.shape)
Y_train = np_utils.to_categorical(y_train, n_classes)
Y_test = np_utils.to_categorical(y_test, n_classes)
print("Shape after one-hot encoding: ", Y_train.shape)

Shape before one-hot encoding:  (60000,)
Shape after one-hot encoding:  (60000, 10)


In [7]:
model = Sequential()
model.add(Conv2D(25, kernel_size=(3,3), strides=(1,1), padding='valid', activation='relu', input_shape=(28,28,1)))
model.add(MaxPool2D(pool_size=(1,1)))
model.add(Flatten())
model.add(Dense(100, activation='relu'))
model.add(Dense(10, activation='softmax'))

model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 25)        250       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 26, 26, 25)       0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 16900)             0         
                                                                 
 dense_5 (Dense)             (None, 100)               1690100   
                                                                 
 dense_6 (Dense)             (None, 10)                1010      
                                                                 
Total params: 1,691,360
Trainable params: 1,691,360
Non-trainable params: 0
____________________________________________

In [8]:
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')
model.fit(X_train, Y_train, batch_size=128, epochs=10, validation_data=(X_test, Y_test))

Epoch 1/10
469/469 [==============================] - 9s 19ms/step - loss: 0.2009 - accuracy: 0.9417 - val_loss: 0.0751 - val_accuracy: 0.9758
Epoch 2/10
469/469 [==============================] - 9s 20ms/step - loss: 0.0614 - accuracy: 0.9818 - val_loss: 0.0761 - val_accuracy: 0.9771
Epoch 3/10
469/469 [==============================] - 10s 20ms/step - loss: 0.0373 - accuracy: 0.9887 - val_loss: 0.0519 - val_accuracy: 0.9839
Epoch 4/10
469/469 [==============================] - 10s 21ms/step - loss: 0.0241 - accuracy: 0.9929 - val_loss: 0.0582 - val_accuracy: 0.9811
Epoch 5/10
469/469 [==============================] - 10s 21ms/step - loss: 0.0167 - accuracy: 0.9952 - val_loss: 0.0549 - val_accuracy: 0.9838
Epoch 6/10
469/469 [==============================] - 10s 21ms/step - loss: 0.0123 - accuracy: 0.9965 - val_loss: 0.0517 - val_accuracy: 0.9852
Epoch 7/10
469/469 [==============================] - 10s 22ms/step - loss: 0.0076 - accuracy: 0.9978 - val_loss: 0.0636 - val_accuracy: 0

#### Using CNN, the validation accuracy is improved to 98%. This is expected since the CNN number of parameters is an order of magnitude larger for the fully-connected network. 